src/mapping/config/MappingConfiguration.cpp

```diff
@@ -24,134 +25,169 @@
+// creates a tag including documentation and stores the tag in the provided
+// \p storage (usually a std::list of tags)
+template <typename Listener, typename TagStorage>
+void createTag(Listener &              listener,
+               const std::string &     name,
+               xml::XMLTag::Occurrence occurrence,
+               const std::string &     xmlNamespace,
+               TagStorage &            storage,
+               std::string             documentation)
+{
+  xml::XMLTag tag(listener, name, occurrence, xmlNamespace);
+  tag.setDocumentation(documentation);
+  storage.push_back(tag);
+}
 
-  XMLTag::Occurrence occ = XMLTag::OCCUR_ARBITRARY;
-  std::list<XMLTag>  tags;
-  {
-    XMLTag tag(*this, VALUE_RBF_TPS, occ, TAG);
-    tag.setDocumentation("Global radial-basis-function mapping based on the thin plate splines.");
-    tags.push_back(tag);
-  }
-  {
-    XMLTag tag(*this, VALUE_RBF_MULTIQUADRICS, occ, TAG);
-    tag.setDocumentation("Global radial-basis-function mapping based on the multiquadrics RBF.");
-    tag.addAttribute(attrShapeParam);
-    tags.push_back(tag);
+// given a list of subtags and parent tags, this function adds all subtags to all
+// parent tags
+void addSubtagsToParents(std::list<xml::XMLTag> &subtags,
+                         std::list<xml::XMLTag> &parents)
+{
+  for (auto &p : parents) {
+    std::for_each(subtags.begin(), subtags.end(), [&p](auto &s) { p.addSubtag(s); });
   }
-  {
-    XMLTag tag(*this, VALUE_RBF_INV_MULTIQUADRICS, occ, TAG);
-    tag.setDocumentation("Global radial-basis-function mapping based on the inverse multiquadrics RBF.");
-    tag.addAttribute(attrShapeParam);
-    tags.push_back(tag);
+}
+
+// this function uses std::variant in order to add attributes of any type (double, string, bool)
+// to all tags in the list of tags \p storage.
+using variant_t = std::variant<xml::XMLAttribute<double>, xml::XMLAttribute<std::string>, xml::XMLAttribute<bool>>;
+template <typename TagStorage>
+void addAttributes(TagStorage &storage, const std::vector<variant_t> &attributes)
+{
+  for (auto &s : storage) {
+    for (auto &a : attributes)
+      std::visit([&s](auto &&arg) { s.addAttribute(arg); }, a);
   }
-  {
-    XMLTag tag(*this, VALUE_RBF_VOLUME_SPLINES, occ, TAG);
-    tag.setDocumentation("Global radial-basis-function mapping based on the volume-splines RBF.");
-    tags.push_back(tag);
+}
+
+// helper for the function below
+template <typename>
+inline constexpr bool always_false_v = false;
+
+// Reads and returns an attribute value of any type (bool, string, double) from the tag,
+// if the attribute is actually contained in the tag. If the attribute is not in the tag,
+// an \p initializationValue is returned.
+template <typename AttributeType>
+AttributeType getAttributeIfPresent(xml::XMLTag &tag, const std::string &attributeName, AttributeType initializationValue)
```

Comment:
With provided default you are referring to the attribute default when adding the attribute to the tag, right?

I agree generally (C++17 enables it now). Do you want me to do this already here? I could imagine that (having the functions here) we create a good first issue out of it and apply the changes in all configuration files?!

Reply:
> With provided default you are referring to the attribute default when adding the attribute to the tag, right?

No, I really mean the default value in case the attribute doesn't exist. Essentially, what you propose in the `getAttributeIfPresent`, which really is more like a `getAttributeWithDefault`.

Reply:
I just realized what you mean here. I'm not an expert enough for `std::optional` in order to pythonize this code. Also, it will kind of erase the advantage of the automatic type deduction for the corresponding type function. I would leave it open and open an issue. Apart from that, this would be something we should change in multiple places, right? Not only affecting the mapping configuration.

Reply:
> I'm not an expert enough for std::optional in order to pythonize this code.

It's actually surprisingly easy to use. (famous last words of a C++ programmer)

```cpp
// Note that "Type" stands for Boolean/Integer/...
XMLTag::getTypeAttributeValue(string name, optional<Type> default_value = std::nullopt)
{
  // ...
  if (found) {
    return value;
  }
  if (default_value) { // checks if optional holds a value, same as .has_value()
    return default_value.value(); // extracts the value.
  }
  PRECICE_ERROR("Not found");
}
```

This is backwards compatible:
```cpp
bool b1 = getBooleanAttributeValue("named");
auto b2 = getBooleanAttributeValue("other", false);
```

> Also, it will kind of erase the advantage of the automatic type deduction for the corresponding type function.

You have to explicitly name the type when extracting the attribute anyhow `getBooleanAttribute("name")`, so there is no deduction taking place.

> Apart from that, this would be something we should change in multiple places, right?

I am not sure where else we have conditional attributes.

Reply:
> You have to explicitly name the type when extracting the attribute anyhow getBooleanAttribute("name"), so there is no deduction taking place.

Right now, you don't have to. Using this (backwards compatible solution), yes.
Anyhow, I will add it then.

Reply:
I see. You refer to the deduction in `getAttributeIfPresent` based on the `initializationValue`.

This indeed, won't work if the `default` parameter is an optional.

An alternative could be an overload.
```cpp
template<class T>
T XMLTag::getAttributeValue(string name); // needs an attribute
template<class T>
T XMLTag::getAttributeValue(string name, T def); // can deduce via the default
```

If it's getting to cumbersome, then feel free stick to your initial solution. This argument lookup needs some refactoring at some point though.

Reply:
Done